### PR TITLE
[MonoVM] Reduce P/Invoke GC transition asserts in release builds

### PR DESCRIPTION
**Pull request by filipnavara**
Profiling shows that large part of the GC transition overhead (~30%) in #58939 is caused by assert-style checks. Disabling them seems to be the best bang-for-the-buck option for reducing the overhead without fundamental changes to the code.

### CODE REVIEW

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/utils/mono-threads-state-machine.c [hunk 1]
```diff
@@ -105,6 +105,7 @@ unwrap_thread_state (MonoThreadInfo* info,
 static void
 check_thread_state (MonoThreadInfo* info)
 {
+#ifdef ENABLE_CHECKED_BUILD_THREAD
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/utils/mono-threads-state-machine.c:108]: Hmm. But to be honest I can't tell the difference - looks like we use both defines for closely related things.

**Review by BrzVlad — APPROVED** [review]

File: src/mono/mono/utils/mono-threads-state-machine.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/mono/mono/utils/mono-threads-state-machine.c:108]: I mostly tried to match the `assert`s in the same file but I can change it. I will update the PR to enable it on debug builds.

[the remark this answers is not in the document]
Comment (filipnavara) [reply src/mono/mono/utils/mono-threads-state-machine.c]: After checking how the configuration options are set I came to the conclusion to change it to the general `ENABLE_CHECKED_BUILD`. That one is defined in all debug builds, the other ones can be selectively enabled.

**Comment by filipnavara** [comment]
On Mac Catalyst ARM64 the test case in #58939 goes from `6,845,245` to `9,126,591` iterations per second, or 33% throughput improvement.

**Review by lambdageek — APPROVED** [review]

**Comment by srxqds** [comment]
will you consider to backport to release/6.0?

**Pull request merged**